File: src/ssh_manager/core/inventory.py
"""Inventory domain models. Keyed by SHA256 fingerprint.

The inventory is the deployment-tracking join that turns rotation into a
checklist. It is version-stamped and persisted atomically (invariant 11).
"""
from __future__ import annotations

from datetime import date, datetime, timedelta
from pathlib import Path

from pydantic import BaseModel, ConfigDict, Field

from ..util import jsonstore
from ..util.errors import ManifestError

SCHEMA_VERSION = 1


class Deployment(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target: str
    method: str                 # ssh-copy-id | web-panel | manual | <adapter>
    date: str | None = None
    verified: bool = False      # False == needs-redeploy


class KeyRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")

    profile: str
    path: str
    type: str = "ed25519"
    comment: str | None = None
    created: str | None = None
    rotate_after_days: int = 365
    expires_on: str | None = None     # derived cache of created + rotate_after_days
    deployments: list[Deployment] = Field(default_factory=list)

    @property
    def needs_redeploy(self) -> bool:
        """True when no deployment is verified (e.g. a freshly minted key)."""
        return not any(d.verified for d in self.deployments)


class Inventory(BaseModel):
    model_config = ConfigDict(extra="forbid")

    version: int = SCHEMA_VERSION
    keys: dict[str, KeyRecord] = Field(default_factory=dict)

    @classmethod
    def load(cls, path: str | Path) -> Inventory:
        path = Path(path)
        if not path.exists():
            return cls()
        try:
            data = jsonstore.read_json(path)
        except ValueError as exc:
            raise ManifestError(f"inventory is not valid JSON: {path}: {exc}") from exc
        except OSError as exc:
            raise ManifestError(f"inventory could not be read: {path}: {exc}") from exc
        try:
            return cls.model_validate(data)
        except Exception as exc:   # pydantic ValidationError -> our typed error
            raise ManifestError(f"inventory failed validation: {path}: {exc}") from exc

    def save(self, path: str | Path) -> None:
        jsonstore.write_json_atomic(path, self.model_dump(mode="json"))

    def record(self, fingerprint: str, rec: KeyRecord) -> None:
        self.keys[fingerprint] = rec


def is_archived_path(path: str) -> bool:
    """True if ``path`` is a rotation's ``/old/`` predecessor slot, i.e.
    ``~/.ssh/profiles/<profile>/old/<key_name>``. Uses the path *structure* (the
    'old' dir sits directly under a profile dir that sits under 'profiles') rather
    than a bare ``"/old/"`` substring, so a profile literally named ``old`` -
    whose active keys live at ``profiles/old/<name>`` - is NOT mistaken for an
    archived key and dropped from expiry/audit."""
    parts = path.split("/")
    # [..., 'profiles', '<profile>', 'old', '<key_name>']
    return len(parts) >= 4 and parts[-2] == "old" and parts[-4] == "profiles"


def compute_expiry(created: str, rotate_after_days: int) -> str:
    """``created`` (YYYY-MM-DD) + rotate_after_days → expires_on (YYYY-MM-DD)."""
    base = datetime.strptime(created, "%Y-%m-%d").date()
    return (base + timedelta(days=rotate_after_days)).isoformat()


def today() -> str:
    return date.today().isoformat()

File: internal/core/inventory/inventory.go
// Package inventory is the deployment-tracking model, ported from
// src/ssh_manager/core/inventory.py. Keyed by SHA256 fingerprint, it turns
// rotation into a checklist and is persisted atomically.
package inventory

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

const schemaVersion = 1

const dateLayout = "2006-01-02"

func decodeStrict(b []byte, v any) error {
	dec := json.NewDecoder(bytes.NewReader(b))
	dec.DisallowUnknownFields()
	return dec.Decode(v)
}

// Deployment records a key installed on one target.
type Deployment struct {
	Target   string  `json:"target"`
	Method   string  `json:"method"` // ssh-copy-id | web-panel | manual | <adapter>
	Date     *string `json:"date"`
	Verified bool    `json:"verified"` // false == needs-redeploy
}

func (d *Deployment) UnmarshalJSON(b []byte) error {
	type alias Deployment
	var aux alias
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*d = Deployment(aux)
	return nil
}

// KeyRecord is one managed key and where it is deployed. All fields serialize
// unconditionally (null for an unset pointer, [] for no deployments) to match
// pydantic's model_dump(mode="json") byte-for-byte.
type KeyRecord struct {
	Profile         string       `json:"profile"`
	Path            string       `json:"path"`
	Type            string       `json:"type"`
	Comment         *string      `json:"comment"`
	Created         *string      `json:"created"`
	RotateAfterDays int          `json:"rotate_after_days"`
	ExpiresOn       *string      `json:"expires_on"`
	Deployments     []Deployment `json:"deployments"`
}

func (r *KeyRecord) UnmarshalJSON(b []byte) error {
	type alias KeyRecord
	aux := alias{Type: "ed25519", RotateAfterDays: 365}
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*r = KeyRecord(aux)
	return nil
}

// MarshalJSON emits a nil Deployments slice as [] (not null), matching pydantic's
// default_factory=list. The pointer fields already emit null when unset.
func (r KeyRecord) MarshalJSON() ([]byte, error) {
	type alias KeyRecord
	a := alias(r)
	if a.Deployments == nil {
		a.Deployments = []Deployment{}
	}
	return json.Marshal(a)
}

// NeedsRedeploy is true when no deployment is verified (e.g. a fresh key).
func (r KeyRecord) NeedsRedeploy() bool {
	for _, d := range r.Deployments {
		if d.Verified {
			return false
		}
	}
	return true
}

// Inventory is the whole inventory, keyed by fingerprint.
type Inventory struct {
	Version int                  `json:"version"`
	Keys    map[string]KeyRecord `json:"keys"`
}

func (inv *Inventory) UnmarshalJSON(b []byte) error {
	type alias Inventory
	aux := alias{Version: schemaVersion, Keys: map[string]KeyRecord{}}
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*inv = Inventory(aux)
	return nil
}

// New returns an empty inventory.
func New() *Inventory { return &Inventory{Version: schemaVersion, Keys: map[string]KeyRecord{}} }

// Load reads an inventory; a missing file yields an empty inventory (matching v1).
func Load(path string) (*Inventory, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return New(), nil
		}
		return nil, fmt.Errorf("inventory could not be read: %s: %w", path, err)
	}
	var inv Inventory
	if err := json.Unmarshal(data, &inv); err != nil {
		return nil, fmt.Errorf("inventory failed validation: %s: %w", path, err)
	}
	if inv.Keys == nil {
		inv.Keys = map[string]KeyRecord{}
	}
	return &inv, nil
}

// Save writes the inventory as indented JSON.
func (inv *Inventory) Save(path string) error {
	b, err := json.MarshalIndent(inv, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, append(b, '\n'), 0o600)
}

// Record stores a key record under its fingerprint.
func (inv *Inventory) Record(fingerprint string, rec KeyRecord) {
	if inv.Keys == nil {
		inv.Keys = map[string]KeyRecord{}
	}
	inv.Keys[fingerprint] = rec
}

// IsArchivedPath reports whether path is a rotation's /old/ predecessor slot,
// i.e. .../profiles/<profile>/old/<key_name> (by structure, so a profile literally
// named "old" is not mistaken for an archived key).
func IsArchivedPath(path string) bool {
	parts := strings.Split(path, "/")
	n := len(parts)
	return n >= 4 && parts[n-2] == "old" && parts[n-4] == "profiles"
}

// ComputeExpiry returns created (YYYY-MM-DD) + rotateAfterDays as YYYY-MM-DD.
func ComputeExpiry(created string, rotateAfterDays int) (string, error) {
	base, err := time.Parse(dateLayout, created)
	if err != nil {
		return "", err
	}
	return base.AddDate(0, 0, rotateAfterDays).Format(dateLayout), nil
}

// Today returns today's date as YYYY-MM-DD.
func Today() string { return time.Now().Format(dateLayout) }
